blueprints/multi-interface-servers.md

```diff
@@ -0,0 +1,829 @@
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+# Additional Server Interfaces
+
+## Problem Description
+Multiple network interfaces may be used by cache servers to service requests.
+However, these cannot be expressed as parts of the current "server" object.
+Specifically, assuming each interface has its own IP address, then each would
+need to be treated as a separate cache server. This means that getting clients
+to use one when the other is unavailable requires manual intervention.
+
+## Proposed Change
+Instead, cache servers should be configurable with numerous network interfaces
+monitored by Traffic Monitor that are all associated with the server, each
+having its own associated network information e.g. IP Address, but Traffic
+Router should only be aware of the "main" interface on which the cache server
+listens for connections.
+
+The specific use-case outlined above, obviously, entails some management be done
+on the cache servers to actually support multiple interfaces answering to the
+same "main" IP address and this should be considered outside of
+<abbr title="Apache Traffic Control">ATC</abbr>'s scope of operation.
+
+## Data Model Impact
+The proposed structure of each of these interfaces is given below as a pair of
+TypeScript<sup>[1](#typescript)</sup> interfaces:
+```typescript
+interface IPAddress {
+	/**
+	 * The actual IP (v4 or v6) address which is used by an interface.
+	 * If it is an IPv6 address
+	 */
+	address: string;
+
+	/**
+	 * The netmask used by the address.
+
+	/**
+	 * The IP (v4 or v6) address of the gateway used by this IP address.
+	 */
+	gateway: string;
+
+	/**
+	 * Tells whether or not this address of this interface is the server's
+	 * "service" address.
+	 * At least one address of EXACTLY ONE interface MUST have this set to
+	 * 'true' for a server.
+	 */
+	serviceAddress: boolean;
+}
+
+interface Interface {
+	/**
+	 * These will be all of the IPv4/IPv6 addresses assigned to the interface,
+	 * including gateways and "masks".
+	 * It is illegal for an interface to not have at least one associated IP
+	 * address.
+	 */
+	ipAddresses: Array<IPAddress> & {0: IPAddress};
+	/**
+	 * The maximum allowed bandwidth for this interface to be considered "healthy"
+	 * by Traffic Monitor.
+	 *
+	 * This has no effect if `monitor` is not true.
+	 * Values are in kb/s.
+	 * The value `0` means "no limit".
+	 */
+	maxBandwidth: bigint;
```

Comment:
Will we continue to discover this from the interface stats?

Reply:
Discover this? This will be set on the object when it's created.

Reply:
That number hasn't been present in the database before as it's derived by the astats plugin to ATS based on which interface name you ask for.

Reply:
There is a minor bug #2500 where it's impossible for astats to figure it out, and a gap where link aggregate member statuses aren't taken into account.

Reply:
Even though it could help, I'm not a fan of moving this to TODB because it's making more state and more opportunity for humans to do it wrong.

Reply:
Oh, I see. This isn't what you guys think it is. This is the equivalent of the `health.threshold.availableBandwidthInKbps` Parameter today.

Reply:
Yes, this is being set in the database so that TM knows what the max_bw is per interface.  We will still use the `health.threshold.kbps` parameter for the overall cache bandwidth threshold.  The idea here is that TM needs to mark a cache unhealth if the overall max bw is reached OR the max bw for an interface is reached.

Reply:
So you want to move the concept of the bandwidth limit partially out of profiles/parameters.  It feels a bit odd to split the implementation like that, but ok.  Instead of fixed caps, could it be either percentage or fixed reservation?  That would allow us to fix the loophole with bonded interfaces in various link state mixes down the road.

Reply:
What loophole?

Also, I think adding support for percentages would entail a different body of work and ought to be considered separately from implementing the multiple-interface servers as outlined in this blueprint.

Reply:
Looking into that, it seems like this really ought to just be solved by having the stats system on the Cache Server read the interface's bonded directory and calculating the totals from its slave interfaces. Which seems like a bug with that/those plugin(s) that should be remedied there regardless of how many interfaces exist on cache server objects in TO.